lib/api/objectCopy.js

```diff
@@ -0,0 +1,288 @@
+                        }
+                    }
+                    return next(null, storeMetadataParams, dataLocator);
+                });
+        },
+        function checkDestAuth(storeMetadataParams, dataLocator, next) {
+            return services.metadataValidateAuthorization(valPutParams,
+                (err, destBucketMD, destObjMD) => {
+                    if (err) {
+                        log.debug('error validating put part of request',
+                        { error: err });
+                        return next(err);
+                    }
+                    const flag = destBucketMD.hasDeletedFlag()
+                        || destBucketMD.hasTransientFlag();
+                    if (flag) {
+                        log.trace('deleted flag or transient flag ' +
+                        'on destination bucket', { flag });
+                        return next(errors.NoSuchBucket);
+                    }
+                    return next(null, storeMetadataParams,
+                        dataLocator, destBucketMD, destObjMD);
+                });
+        },
+        function getCipherBundle(storeMetadataParams, dataLocator, destBucketMD,
+            destObjMD, next) {
+            const serverSideEncryption = destBucketMD.getServerSideEncryption();
+            if (!serverSideEncryption) {
+                return next(null, storeMetadataParams, dataLocator,
+                    destObjMD, null);
+            }
+            return kms.createCipherBundle(
+                serverSideEncryption,
+                log, (err, cipherBundle) => {
+                    if (err) {
+                        log.debug('error getting cipherBundle');
+                        return next(errors.InternalError);
+                    }
+                    return next(null, storeMetadataParams, dataLocator,
+                        destObjMD, cipherBundle);
+                });
+        },
+        function goGetData(storeMetadataParams, dataLocator,
+            destObjMD, cipherBundle, next) {
+            // skip if source and dest the same or 0 byte object
+            if (sourceIsDestination || dataLocator.length === 0) {
+                return next(null, storeMetadataParams, dataLocator, destObjMD,
+                    cipherBundle);
+            }
+             // dataLocator is an array.  need to get and put all parts
+             // For now, copy 1 part at a time. Could increase the second
+             // argument here to increase the number of parts
+             // copied at once.
+            return async.mapLimit(dataLocator, 1,
+                // eslint-disable-next-line prefer-arrow-callback
+                function copyPart(part, cb) {
+                    return data.get(part, log, (err, stream) => {
+                        if (err) {
+                            return cb(err);
+                        }
+                        return data.put(cipherBundle, stream, part.size,
```

Comment:
as a consequence of this, cipheredDataKey from the cipher bundle should be kept per part and registered in the destination's location array